File: src/main/java/com/morak/performancetracker/aop/persistence/QueryMonitor.java
package com.morak.performancetracker.aop.persistence;

import com.morak.performancetracker.Monitor;
import org.springframework.stereotype.Component;

@Component
public class QueryMonitor implements Monitor {

    private String query;
    private long startTime;
    private double queryTime;

    public void start() {
        startTime = System.nanoTime();
    }

    public void end() {
        if (startTime == 0) {
            throw new IllegalStateException("Monitor not initialized yet");
        }
        this.queryTime += System.nanoTime() - startTime;
    }

    @Override
    public void clear() {
        this.query = null;
        this.startTime = 0;
        this.queryTime = 0;
    }

    @Override
    public String getSignature() {
        return "Query";
    }

    public long getStartTime() {
        return startTime;
    }

    public double getQueryTime() {
        return queryTime;
    }

    public String getQuery() {
        return query;
    }

    public void setQuery(String query) {
        this.query = query;
    }
}

File: src/main/kotlin/com/morak/performancetracker/aop/persistence/QueryMonitor.kt
package com.morak.performancetracker.aop.persistence

import com.morak.performancetracker.Monitor
import org.springframework.stereotype.Component

@Component
class QueryMonitor(
    var query: String? = null,
) : Monitor {

    var startTime: Long = 0
    var queryTime = 0.0

    override fun clear() {
        query = null
        startTime = 0
        queryTime = 0.0
    }

    override val signature: String
        get() = "Query"

    fun start() {
        startTime = System.nanoTime()
    }

    fun end() {
        check(startTime != 0L) { "Monitor not initialized yet" }
        queryTime += (System.nanoTime() - startTime).toDouble()
    }
}
